src/datasets.jl

```
@@ -0,0 +1,236 @@
+struct Dataset{K, T<:XArray}
+    # Dims are restricted to symbols because that's a NamedDims requirement
+    dims::OrderedSet{Symbol}
+    # Data lookup can be by any type, but typically it'll either be symbol or tuple.
+    data::LittleDict{K, T}
+end
+
+# Taking pairs is the most general constructor as it doesn't make assumptions about the
+# data key type.
+function Dataset(pairs::Pair{K}...; dims=()) where K
+    data = LittleDict{K, XArray}(pairs...)
+    dimset = OrderedSet(isempty(dims) ? Iterators.flatten(dimnames.(values(data))) : dims)
+
+    result = Dataset(dimset, data)
+    _isvalid(result) || throw(ArgumentError("Shared axes don't match."))
+    return result
+end
+
+# dataset(table, dims::Symbol...) -> Dataset(dims, arrays for each remaining value columns)
+function Dataset(table; dims=())
+    data = LittleDict{Symbol, XArray}()
+    # We need to extrac the columns to so we can isolate all values columns
+    cols = Tables.columns(table)
+    valcols = setdiff(Tables.columnnames(cols), dims)
+
+    # It'd be faster to just make a copy after the first value column since the axes
+    # shouldn't be changing
+    for c in valcols
+        data[c] = wrapdims(table, c, dims...; default=missing, sort=true, force=true)
+    end
+
+    dimset = OrderedSet(isempty(dims) ? Iterators.flatten(dimnames.(values(data))) : dims)
+    result = Dataset(dimset, data)
+
+    # This shouldn't be necessary because we're using the same axes for all components.
+    _isvalid(result) || throw(ArgumentError("Shared axes don't match."))
```

would it be better to throw the error inside `_isvalid`? Seems like we always run both commands together.

---

Yeah, I was initially thinking I'd be able to handle these cases differently depending on context, but I think having the same simple error conditions is probably better now.